classes/mail/Mailable.inc.php

```diff
@@ -0,0 +1,300 @@
+                QueuedPayment::class => QueuedPaymentEmailVariable::class,
+            ];
+    }
+
+    /**
+     * @param array $users
+     * @return bool
+     */
+    protected static function arrayOfUsers(array $users) : bool
+    {
+        foreach ($users as $user) {
+            if (!is_a($user, User::class))
+                return false;
+        }
+        return true;
+    }
+
+    /**
+     * Arguments are scanned to retrieve variables which can be assigned to the template of the email
+     * @param array $args
+     */
+    protected function setTemplateVariables(array $args): void
+    {
+        $map = static::templateVariablesMap();
+        foreach ($args as $arg) {
+            // Treat array as recipients if it contains only User objects
+            if (is_array($arg) && static::arrayOfUsers($arg)) {
+                $recipientsVars = new $map[User::class]['recipients']($arg);
+                $this->viewData = array_merge(
+                    $this->viewData,
+                    $recipientsVars->getValue()
+                );
+                continue;
+            }
+
+            // Treat User object passed to the constructor as sender
+            if (is_a($arg, User::class)) {
+                $senderVars = new $map[User::class]['sender']($arg);
+                $this->viewData = array_merge(
+                    $this->viewData,
+                    $senderVars->getValue()
+                );
+                continue;
+            }
+
+            foreach ($map as $className => $assoc) {
+                if (is_a($arg, $className)) {
+                    $assocVariable = new $assoc($arg);
+                    $this->viewData = array_merge(
+                        $this->viewData,
+                        $assocVariable->getValue()
+                    );
+                    continue 2;
+                }
+            }
+
+            // Give up, object isn't mapped
+            $type = is_object($arg) ? get_class($arg) : gettype($arg);
+            throw new InvalidArgumentException($type . ' argument passed to the ' . static::class . ' constructor isn\'t associated with template variables');
+        }
+    }
+
+    /**
+     * Retrieves array of variables that can be assigned to email templates
+     * @return array ['variableName' => description]
+     */
+    public static function getTemplateVarsDescription() : array
+    {
+        $args = static::getParamsClass(static::getConstructor());
+        $map = static::templateVariablesMap();
+        $variables = [];
+        foreach ($args as $arg) { /** @var  ReflectionParameter $arg) */
+            $class = $arg->getType()->getName();
+
+            if ($class === 'array') {
+                $variables = array_merge(
+                    $variables,
+                    $map[User::class]['recipients']::getDescription()
+                );
+                continue;
+            }
+
+            if (!array_key_exists($class, $map)) continue;
```

Comment:
Let's use the `{` block here. It's just personal preference, but I find it easy to miss when the `continue` statement is all the way over on the right.